neps/nep-0393.md

```diff
@@ -0,0 +1,480 @@
+
+If an SBT smart contract doesn't opt-in to any registry, then we should think about it as a single token registry, and it MUST strictly implement all SBT Registry query functions by itself: the contract address must be part of the arguments, and it must check that it equals to the deployed account address (`require!(ctr == env::current_account_id())`).
+
+Moreover, a registry will provide an efficient way to query multiple tokens for a single user. This will allow implementation of use cases such us:
+
+- SBT based identities (main use case of the `i-am-human` protocol);
+- SBT classes;
+- decentralized societies.
+
+#### Recovery within an SBT Registry
+
+SBT registry can define it's own mechanism to atomically recover all tokens related to one account and execute soul transfer to another account, without going one by one through each SBT issuer (sometimes that might be even not possible). Example mechanisms a Registry can implement to recovery mechanism:
+
+- KYC based recovery
+- Social recovery
+
+![Social Recovery, Image via “Decentralized Society”](https://bankless.ghost.io/content/images/public/images/cdb1fc23-6179-44f0-9bfe-e5e5831492f7_1399x680.png)
+
+SBT Registry based recovery is not part of this specification.
+
+#### Minting authorization
+
+A registry can limit which contracts can mint SBTs by implementing a custom issuer registration methods. Example: a simple access control list managed by a DAO.
+
+### Smart contract interface
+
+For the Token ID type we propose `u64` rather than `U128`. `u64` capacity is more than 1e19. If we will mint 10'000 SBTs per second, then it will take us 58'494'241 years to fill the capacity.
+Today, the JS integer limit is `2^53-1 ~ 9e15`. Similarly, when minting 10'000 SBTs per second, it will take us 28'561 years to reach the limit. So, we don't need u128 nor the String type. However, if for some reason, we will need to get u64 support for JS, then we can always add another set of methods which will return String, so making it compatible with NFT standard (which is using `U128`, which is a string). Also, it's worth to note, that in 28'000 years JS (if it will still exists) will be completely different.
+
+```rust
+// TokenId and ClassId must be positive (0 is not a valid ID)
+pub type TokenId = u64;
+pub type ClassId = u64;
+
+pub struct Token {
+    pub token: TokenId,
+    pub owner: AccountId,
+    pub metadata: TokenMetadata,
+}
+```
+
+The Soulbound Token follows the NFT [NEP-171](https://github.com/near/NEPs/blob/master/neps/nep-0171.md) interface, with few differences:
+
+- token ID is `u64` (as discussed above).
+- token class is `u64`, it's required when minting and it's part of the token metadata.
+- `TokenMetadata` doesn't have `title`, `description`, `media`, `media_hash`, `copies`, `extra`, `starts_at` nor `updated_at`. All that attributes except the `updated_at` can be part of the document stored at `reference`. `updated_at` can be tracked easily by indexers.
```

Comment:
> `updated_at` can be tracked easily by indexers.

This is not correct. Method responses (Such as the response to `sbt()`) are not easy to track by indexers, as any TX to the contract may result in a change to any or all tokens.
The way NFTs solve this, using NEP-297, is by emitting a `contract_metadata_update`. Which is missing from this proposal.

Further reading: [NFT Events](https://nomicon.io/Standards/Tokens/NonFungibleToken/Event)

Reply:
> keep track about updates, now how to listen on method response.

Indexers can listen to:
 - Method calls and their arguments, and responses
 - Events

Surely, with the wide range of options, many kind of method calls can cause token values to update. Which is not in the standard, and thus you have no reasonable expectation that indexers can implement each different registries.

Thus, you need to tell indexers that some metadata has changed. So you need the event that tells the indexers just that. They can then make a call and pull the updated metadata.

Reply:
@alexastrum actually there is `trait SBTContract` (the Token interface) in the reference implementation, but was not copied here. Thanks more mentioning it, I will add a section about it.

Reply:
I raised it on this line, because you make a broad claim, without backing it up. The line 2 bellow just says you have events for it.
What I am highlighting is that your events are lacking. It does not emit for token updates. So there is no way to track that.
I'm happy to raise it again at the events if that is your concern. But the point stands. You do not have an update event, which lets indexers know a token has changed.

Reply:
@robert-zaremba Maybe just remove that wording about tracking of `updated_at`, because it's currently not true according to the current NEP iteration: there is no methods nor events about metadata change (just expiration date can be tracked indirectly as a specific case of your API/events).